azure-keyvault-cryptography/src/main/java/com/microsoft/azure/keyvault/cryptography/algorithms/RsaSignature.java

```diff
@@ -10,209 +10,414 @@
-    	BigInteger e = K.getPublicExponent();
-    	
-    	if ( s.compareTo(BigInteger.ONE) == -1 || s.compareTo(n) != -1 ) {
-    		throw new IllegalArgumentException("message representative out of range");
-    	}
-    	
-    	return s.modPow(e, n);
+
+        if (K == null) {
+            throw new IllegalArgumentException("K");
+        }
+
+        if (s == null) {
+            throw new IllegalArgumentException("s");
+        }
+        BigInteger n = K.getModulus();
+        BigInteger e = K.getPublicExponent();
+
+        if (s.compareTo(BigInteger.ONE) == -1 || s.compareTo(n) != -1) {
+            throw new IllegalArgumentException("message representative out of range");
+        }
+
+        return s.modPow(e, n);
+    }
+    
+    /**
+     * See https://tools.ietf.org/html/rfc3447#section-9.1.1
+     * 
+     * @param mHash the hashed message (hashed with the corresponding digest: i.e., SHA-256 for PS256, SHA-384 for PS384, SHA-512 for PS512 
+     * @param emBits the maximal bit length of the integer OS2IP(EM)
+     * @param messageDigest the messageDigest (or hash) that corresponds to the algorithm
+     * @return
+     */
+    protected byte[] EMSA_PSS_ENCODE_HASH(byte[] mHash, int emBits, MessageDigest messageDigest) {
+
+        // Let mHash = Hash(M), an octet string of length hLen.
+        // function takes in the mHash.
+
+        // salt length for the service corresponds to the digest length, which is the hash length / 8
+        int sLen = messageDigest.getDigestLength();
+        int hLen = mHash.length;
+        int emLen = (int) Math.ceil(emBits / 8.0);
+
+        // If emLen < hLen + sLen + 2, output "encoding error" and stop.
+        if (emLen < (hLen + sLen + 2)) {
+            throw new IllegalArgumentException("encoding error");
+        }
+
+        // Generate a random octet string salt of length sLen; if sLen = 0,
+        // then salt is the empty string.
+
+        byte[] salt = new byte[sLen];
+        SecureRandom rng = new SecureRandom();
+        rng.nextBytes(salt);
+
+        // 5. Let
+        // M' = (0x)00 00 00 00 00 00 00 00 || mHash || salt;
+        // M' is an octet string of length 8 + hLen + sLen with eight
+        // initial zero octets.
+        messageDigest.update(new byte[8]);
+        messageDigest.update(mHash);
+        messageDigest.update(salt);
+
+        // 6. Let H = Hash(M'), an octet string of length hLen.
+        byte[] H = messageDigest.digest();
+
+        // 7. Generate an octet string PS consisting of emLen - sLen - hLen - 2
+        // zero octets. The length of PS may be 0.
+        byte[] PS = new byte[emLen - sLen - hLen - 2];
+
+        // 8. Let DB = PS || 0x01 || salt; DB is an octet string of length
+        // emLen - hLen - 1.
+
+        byte[] DB = new byte[emLen - hLen - 1];
+
+        System.arraycopy(PS, 0, DB, 0, emLen - sLen - hLen - 2);
+        DB[emLen - sLen - hLen - 2] = (byte) 0x01;
+        System.arraycopy(salt, 0, DB, emLen - sLen - hLen - 1, sLen);
```

Comment:
This would probably be more clear if you precalc the length of PS and put it in a variable or use PS.lenght()